src/vs/workbench/electron-browser/actions.ts

```diff
@@ -918,4 +922,289 @@ export class ToggleSharedProcessAction extends Action {
+		@IEditorGroupService groupService: IEditorGroupService,
+		@IPanelService panelService: IPanelService,
+		@IPartService partService: IPartService,
+		@IViewletService viewletService: IViewletService
+	) {
+		super(id, label, editorService, groupService, panelService, partService, viewletService);
+	}
+
+	protected navigateOnEditorFocus(isEditorGroupVertical, isSidebarPositionLeft): TPromise<boolean> {
+		if (isEditorGroupVertical) {
+			return this.navigateRightFromEditor(true, isSidebarPositionLeft);
+		}
+		return this.navigateRightFromEditor(false, isSidebarPositionLeft);
+	}
+
+	protected navigateOnPanelFocus(isEditorGroupVertical, isSidebarPositionLeft): TPromise<boolean> {
+		if (!isSidebarPositionLeft) {
+			return this.navigateToSidebar();
+		}
+		return TPromise.as(false);
+	}
+
+	protected navigateOnSidebarFocus(isEditorGroupVertical, isSidebarPositionLeft): TPromise<boolean> {
+		if (!isSidebarPositionLeft) {
+			return TPromise.as(false);
+		}
+		if (isEditorGroupVertical) {
+			return this.navigateToStartOfEditor();
+		}
+		return this.navigateToLastActiveEditor();
+	}
+
+	private navigateRightFromEditor(jumpGroups: boolean, isSidebarPositionLeft: boolean): TPromise<boolean> {
+		const model = this.groupService.getStacksModel();
+		let result = model.next(jumpGroups, false);
+		if (result) {
+			return this.editorService.openEditor(result.editor, null, model.positionOfGroup(result.group))
+				.then(ok => true);
+		}
+		if (!isSidebarPositionLeft) {
+			return this.navigateToSidebar();
+		}
+		return TPromise.as(false);
+	}
+}
+
+export class NavigateUpAction extends BaseNavigationAction {
+
+	public static ID = 'workbench.action.navigateUp';
+	public static LABEL = nls.localize('navigateUp', "Navigate Up");
+
+	constructor(
+		id: string,
+		label: string,
+		@IWorkbenchEditorService editorService: IWorkbenchEditorService,
+		@IEditorGroupService groupService: IEditorGroupService,
+		@IPanelService panelService: IPanelService,
+		@IPartService partService: IPartService,
+		@IViewletService viewletService: IViewletService
+	) {
+		super(id, label, editorService, groupService, panelService, partService, viewletService);
+	}
+
+	protected navigateOnEditorFocus(isEditorGroupVertical, isSidebarPositionLeft): TPromise<boolean> {
+		if (isEditorGroupVertical) {
+			return TPromise.as(false);
+		}
+		return this.navigateUpFromEditor();
+	}
+
+	protected navigateOnPanelFocus(isEditorGroupVertical, isSidebarPositionLeft): TPromise<boolean> {
+		if (isEditorGroupVertical) {
+			return this.navigateToLastActiveEditor();
+		}
+		return this.navigateToEndOfEditor();
+	}
+
+	private navigateUpFromEditor(): TPromise<boolean> {
+		const model = this.groupService.getStacksModel();
+		const position = model.positionOfGroup(model.activeGroup);
+		this.groupService.focusGroup(position - 1);
+		return TPromise.as(true);
+	}
+}
+
+export class NavigateDownAction extends BaseNavigationAction {
```

Comment:
It's my understanding that we will not move down or up when the focus is in the sidebar.
However, if we should navigate down to the panel when focus is in the sidebar, it can be implemented.
Should it be implemented?